metriq_gym/helpers/graph_helpers.py

```diff
@@ -57,20 +61,68 @@ def device_graph_coloring(topology_graph: rx.PyGraph) -> GraphColoring:
     that can be executed without interference. These pairs are grouped by "color." The coloring reduces
     the complexity of the benchmarking process by organizing the graph into independent sets of qubit pairs.
 
+    Chooses between:
+      - One-factorization (optimal) for complete graphs,
+      - Bipartite edge-coloring for bipartite graphs,
+      - Greedy edge-coloring otherwise.
+
     Args:
         topology_graph: The topology graph (coupling map) of the quantum device.
 
     Returns:
         GraphColoring: An object containing the coloring information.
     """
     num_nodes = topology_graph.num_nodes()
+    num_edges = topology_graph.num_edges()
 
-    # Graphs are bipartite, so use that feature to prevent extra colors from greedy search.
-    # This graph is colored using a bipartite edge-coloring algorithm.
-    edge_color_map = rx.graph_bipartite_edge_color(topology_graph)
+    # Use optimal one-factorization if graph is complete
+    if num_edges == num_nodes * (num_nodes - 1) // 2:
+        edge_color_map = _complete_graph_edge_color(topology_graph)
+    else:
+        try:
+            edge_color_map = rx.graph_bipartite_edge_color(topology_graph)
+        except rx.GraphNotBipartite:
+            edge_color_map = rx.graph_greedy_edge_color(topology_graph)
 
-    # Get the index of the edges.
     edge_index_map = dict(topology_graph.edge_index_map())
     return GraphColoring(
-        num_nodes=num_nodes, edge_color_map=edge_color_map, edge_index_map=edge_index_map
+        num_nodes=num_nodes,
+        edge_color_map=edge_color_map,
+        edge_index_map=edge_index_map,
     )
+
+
+def _complete_graph_edge_color(topology_graph: rx.PyGraph) -> dict[int, int]:
```

Comment:
If we add something like this (no matter whether we sneak a `_` in the prefix :D), we need to add tests for it

Reply:
Good point. Done in 6d881e697d18494dc50bfe1013223f2cd0658441